staging/kos/pkg/controllers/connectionagent/connection_agent.go

```diff
@@ -0,0 +1,1114 @@
+	remoteAttsInformer k8scache.SharedIndexInformer
+
+	// Channel to close to stop remoteAttsInformer when the virtual network
+	// becomes irrelevant.
+	remoteAttsInformerStopCh chan struct{}
+}
+
+// layer2VirtualNetworksState is the second layer of state associated with all
+// the relevant virtual networks. Its main purposes are set up of layer1VNState
+// when a virtual network becomes relevant and clearing such layer1VNState when
+// the associated virtual network becomes irrelevant (and this in turn triggers
+// deletion of the network interfaces of remote NetworkAttachments in that
+// virtual network).
+// All operations on a layer2VirtualNetworksState while queue workers are
+// running must be done with the mutex locked.
+type layer2VirtualNetworksState struct {
+	sync.Mutex
+
+	// localAttToLayer2VNI maps a local NetworkAttachment namespaced name to
+	// the VNI of the layer2VirtualNetworkState where it's stored.
+	localAttToLayer2VNI map[k8stypes.NamespacedName]uint32
+
+	// vniToVNState maps a VNI to its layer2VirtualNetworkState.
+	vniToVNState map[uint32]*layer2VirtualNetworkState
+}
+
+// ConnectionAgent represents a K8S controller which runs on every node of the
+// cluster and eagerly maintains up-to-date the mapping between virtual IPs and
+// host IPs for every relevant NetworkAttachment. A NetworkAttachment is
+// relevant to a connection agent if: (1) it runs on the same node as the
+// connection agent, or (2) it's in a virtual network where at least one
+// NetworkAttachment for which (1) is true exists. To achieve its goal, a
+// connection agent receives notifications about relevant NetworkAttachments
+// through Informers, and when necessary creates/updates/deletes network
+// interfaces through a low-level network interface fabric. When a new virtual
+// network becomes relevant, a new informer on remote NetworkAttachments in that
+// virtual network is created. Upon being notified of the creation of a local
+// NetworkAttachment, the connection agent also updates the status of such
+// attachment with its host IP and the name and the MAC address of the interface
+// which was created.
+type ConnectionAgent struct {
+	node          string
+	hostIP        gonet.IP
+	kcs           *kosclientset.Clientset
+	netv1a1Ifc    netvifc1a1.NetworkV1alpha1Interface
+	eventRecorder k8seventrecord.EventRecorder
+	queue         k8sworkqueue.RateLimitingInterface
+	workers       int
+	netFabric     netfabric.Interface
+	stopCh        <-chan struct{}
+
+	// Informer and lister on NetworkAttachments on the same node as the
+	// connection agent.
+	localAttsInformer k8scache.SharedIndexInformer
+	localAttsLister   koslisterv1a1.NetworkAttachmentLister
+
+	// Layer 1 of the state associated with all relevant virtual networks.
+	// Always access while holding its mutex.
+	// Never attempt to lock l2VirtNetsState's mutex while holding
+	// l1VirtNetsState's, it can lead to deadlock.
```

Comment:
There's only one l2VirtNetsState's mutex and there's only one l1VirtNetsState's mutex.

Reply:
There's only one object of type `layer1VirtualNetworksState` (plural of virtual network; it embeds a mutex) and many objects of `layer1VirtualNetworkState`(singular of virtual network; there's no mutex). The same applies for layer 2. Admittedly though it might be easy to miss because the difference between the type names is only an `s` in the middle of the names. Regardless of the layer, the objects with the plural names (aggregate objects) represent the state that pertains to all virtual networks, while each object with singular name represents the state that pertains to a single virtual network. Notice that the aggregate objects also contain all the per-virtual network objects in a map from VNI to virtual network state for that VNI. 

The fact that only aggregate objects embed a mutex limits parallelism: in some parts of the code, processing of per-virtual network state objects of different virtual networks must happen in mutual exclusion even if conceptually it does not have to, because it must happen while holding the aggregate object's mutex. I don't remember the details that led to this choice, but I remember that I thought about alternatives and discarded them on the basis that they made the code more complicated. I'm not saying we can't do that, but it's not as straightforward and simple as it might look at a first glance (one thing is that if you keep a map from VNI  to per-virtual network state for that VNI you still need a global mutex to access that map, but I remember there were other issues as well).

Reply:
Oh, you're right, there are just those two mutexes.

Regarding the limited parallelism, it's not bad.  The work done while holding either lock is limited to local data twiddling, no remote operations are done.  I think we should live with that unless and until we have quantitative evidence that it's a problem.

Reply:
Agreed.